LayoutTests/accessibility/ax-thread-text-apis/text-marker-word-nav.html

```diff
@@ -0,0 +1,191 @@
+
+<div id="text2">
+c <img src="#" aria-label="blah" style="background-color: #aaaaaa; width: 100px; height: 100px;">d
+</div>
+
+<div class="userselect" id="text3">can't select</div>
+
+<div id="text4">
+巧克力是食物吗？
+</div>
+<div id="text4a">
+كيف حالك؟
+</div>
+
+<pre id="text5">
+both   spaces
+line breaks
+</pre>
+
+<div id="text6">
+some<br>text
+</div>
+
+<p id="text8">
+<strong>Edit</strong>
+text End
+</p>
+
+
+<script>
+var output = "This tests that word navigation is working correctly.\n";
+
+if (window.accessibilityController) {
+    var text = accessibilityController.accessibleElementById("text");
+    // Get the actual text node.
+    text = text.childAtIndex(0);
+
+    // Check that we can get the word range. Land at "w" in "word1 test".
+    var textMarkerRange = text.textMarkerRangeForElement(text);
+    var startMarker = text.startTextMarkerForTextMarkerRange(textMarkerRange);
+    var currentMarker = advanceAndVerify(startMarker, 1, text);
+    
+    // Check that we are at the end of paragraph, so right word should be empty
+    currentMarker = advanceAndVerify(currentMarker, 9, text);
+    
+    // Check the case with span
+    // At "T" in "Thisis", should return the word as "Thisislongword".
+    currentMarker = advanceAndVerify(currentMarker, 1, text);
+    // // At " " before "I", the word should be "I'll".
+    currentMarker = advanceAndVerify(currentMarker, 14, text);
+    // // At " " before "try", the word should excludes "."
+    currentMarker = advanceAndVerify(currentMarker, 5, text);
+
+    // Check the case with contenteditable
+    // At "e" in "editable", the word should NOT include "Content" before it.
+    currentMarker = advanceAndVerify(currentMarker, 18, text);
+
+    // // Check the case with replaced node, the replaced node should be considered a word.
+    var text2 = accessibilityController.accessibleElementById("text2");
+    textMarkerRange = text2.textMarkerRangeForElement(text2);
+    currentMarker = text2.startTextMarkerForTextMarkerRange(textMarkerRange);
+    currentMarker = advanceAndVerify(currentMarker, 2, text2);
+    currentMarker = advanceAndVerify(currentMarker, 1, text2);
+
+    // Check user-select:none is also working.
+    var text3 = accessibilityController.accessibleElementById("text3");
+    textMarkerRange = text3.textMarkerRangeForElement(text3);
+    currentMarker = text3.startTextMarkerForTextMarkerRange(textMarkerRange);
+    currentMarker = advanceAndVerify(currentMarker, 1, text3);
+
+    // Check multi-language, Chinese here.
+    var text4 = accessibilityController.accessibleElementById("text4");
+    textMarkerRange = text4.textMarkerRangeForElement(text4);
+    currentMarker = text4.startTextMarkerForTextMarkerRange(textMarkerRange);
+    // Make sure when we are at the beginning of line, it won't go to previous node.
+    // FIXME: Left word in the live tree returns a word, which is inconsistent with other left words when at the beginning of a word.
```

Comment:
Do we know why the main-thread text marker implementation behaves differently / inconsistently in this case?

Reply:
I haven't been able to nail this down precisely. The main thread/visible position has a lot of code to handle stopping at word boundaries. `<br>` characters are handled very differently in several scenarios.